File: Backend/src/main/java/com/neuroforge/service/SpecService.java
package com.neuroforge.service;

import com.neuroforge.dto.spec.SpecRequest;
import com.neuroforge.dto.spec.SpecResponse;
import com.neuroforge.dto.spec.SpecSummaryResponse;
import com.neuroforge.dto.spec.SpecVersionDTO;

import java.util.List;

public interface SpecService {
    List<SpecSummaryResponse> getSpecsByProject(Long projectId, String username);
    SpecResponse getSpecById(Long specId, String username);
    SpecResponse createSpec(Long projectId, SpecRequest request, String username);
    SpecResponse updateSpec(Long specId, SpecRequest request, String username);
    List<SpecVersionDTO> getSpecVersions(Long specId, String username);
    SpecResponse getSpecByVersion(Long specId, Integer version, String username);
    SpecResponse submitForReview(Long specId, String username);
    SpecResponse approveSpec(Long specId, String username);
    SpecResponse requestChanges(Long specId, String note, String username);
}

File: Backend/frontend/src/services/spec.service.ts
import api from '../lib/api';

export interface UserStoryDTO {
  id: string;
  asA: string;
  iWant: string;
  soThat: string;
  criteria: string[];
}

export interface SpecVersionDTO {
  version: number;
  status: string;
  updatedAt: string;
  updatedBy?: string;
}

export interface SpecResponse {
  id: number;
  title: string;
  description: string;
  status: string;
  version: number;
  userStories: UserStoryDTO[];
  functionalRequirements: string[];
  nonFunctionalRequirements: string[];
  versions: SpecVersionDTO[];
}

export interface SpecSummaryResponse {
  id: number;
  title: string;
  storyCount: number;
  updatedAt: string;
  version: number;
  status: string;
}

export interface SpecRequest {
  title: string;
  description: string;
  userStories?: UserStoryDTO[];
  functionalRequirements?: string[];
  nonFunctionalRequirements?: string[];
}

export const specService = {
  getByProject: (projectId: number) =>
    api.get<SpecSummaryResponse[]>(`/projects/${projectId}/specs`),

  create: (projectId: number, data: SpecRequest) =>
    api.post<SpecResponse>(`/projects/${projectId}/specs`, data),

  getById: (specId: number) =>
    api.get<SpecResponse>(`/specs/${specId}`),

  update: (specId: number, data: SpecRequest) =>
    api.put<SpecResponse>(`/specs/${specId}`, data),

  getVersions: (specId: number) =>
    api.get<SpecVersionDTO[]>(`/specs/${specId}/versions`),

  getVersion: (specId: number, version: number) =>
    api.get<SpecVersionDTO>(`/specs/${specId}/versions/${version}`),

  submit: (specId: number) =>
    api.post<SpecResponse>(`/specs/${specId}/submit`),

  approve: (specId: number) =>
    api.post<SpecResponse>(`/specs/${specId}/approve`),

  requestChanges: (specId: number, data: { note: string }) =>
    api.post<SpecResponse>(`/specs/${specId}/request-changes`, data),
};
